[JENKINS-29829] Stabilize and test on various HW/SW platforms

pjanouse opened the pull request:
Main changed in the code and test-suite:
- Added dependency to used findbugs-annotation
- Added decisions making based on current support state of JNA, implemented fall-back mechanism if possible
- Proper plug-in work on various HW/SW platforms (where didn't work correctly previously)
- Don't eat but throw further exceptions like IOException, InterruptedException
- Unix like systems divided into two categories:
  -- supported by JNA (if supported and verified that it works)
  -- unsupported by JNA or tested that doesn't work yet - used fall-back mechanism via CLI execution (kill command)
- Previous item includes renaming of several impl. classes for better understanding
- Used own JenkinsRule and necessary dependencies (JenkinsRecipe, JenkinsComputerConnectorTester) until JENKINS-29507 fixed
- Explicitly closing InputStreams after cmd execution - prevention of stucking on some platforms (like MS Windows 2003)

Known limitations/issues:
- execution of test-suite on MS Windows 2008 ia64 arch doesn't work due to JNA unsupporting that platform (failing in Jenkins Core)

pjanouse committed:
[JENKINS-29829] Stabilize and test on various HW/SW platforms

Main changed in the code and test-suite:
- Added dependency to used findbugs-annotation
- Added decisions making based on current support state of JNA, implemented fall-back mechanism if possible
- Proper plug-in work on various HW/SW platforms (where didn't work correctly previously)
- Don't eat but throw further exceptions like IOException, InterruptedException
- Unix like systems divided into two categories:
  -- supported by JNA (if supported and verified that it works)
  -- unsupported by JNA or tested that doesn't work yet - used fall-back mechanism via CLI execution (kill command)
- Previous item includes renaming of several impl. classes for better understanding
- Used own JenkinsRule and necessary dependencies (JenkinsRecipe, JenkinsComputerConnectorTester) until JENKINS-29507 fixed
- Explicitly closing InputStreams after cmd execution - prevention of stucking on some platforms (like MS Windows 2003)

Known limitations/issues:
- execution of test-suite on MS Windows 2008 ia64 arch doesn't work due to JNA unsupporting that platform (failing in Jenkins Core)

vjuranek commented on src/test/java/org/jvnet/hudson/test/MyJenkinsRule.java:
```diff
@@ -0,0 +1,2177 @@
+ * @author Stephen Connolly
+ * @since 1.436
+ * @see RestartableJenkinsRule
+ */
```
If possible, I'd prefer test dependency on artifact which originally implement it and not copy-paste the code.
If not possible, could you shortly describe why `JenkinsRule` is not sufficient?
Thanks

vjuranek replied:
IMHO link to JENKINS-29507 would be sufficient in the later case

pjanouse replied:
I'd prefer the same approach and not copy-paste almost the same code, but I can't override a static initialization block in Java code which causes an issue (details are in [1])

Because I need to avoid an execution of a static initialization block of JenkinsRule in the testsuite, I have to remove any code which references org.jvnet.hudson.test.JenkinsRule, including an interface(s). For that reason I have to implement all three classes (MyJenkinsRule, MyJenkinsRecipe, MyJenkinsComputerConnectorTester) by myself again.

If you have an idea how to solve the issue by other way, I'm open to discuss it again.

[1] https://issues.jenkins-ci.org/browse/JENKINS-29507

pjanouse committed:
Solaris 9 removed from JNA compatible systems

Reason: 'java.lang.UnsatisfiedLinkError: Error looking up function 'unsetenv': ld.so.1: java: fatal: unsetenv: can't find symbol'

vjuranek commented:
Merged, thanks!
